### PR TITLE
Refactor EVM modules, impl comments suggestions 

### PR DESCRIPTION
**Pull request by Mr-Leshiy**
Implement suggestion from this PR https://github.com/input-output-hk/chain-libs/pull/625

### CODE REVIEW

**Review by enzoc4 — APPROVED** [review]
I'm lacking context as I haven't been involved with the EVM changes, but code looks good to me.

I wonder whether it was possible to use the blake2 implementation from cryptoxide that we're using in `chain-crypto`, but it's unrelated to this PR and maybe already discussed.

File: chain-evm/src/precompiles/blake2.rs [hunk 1]
```diff
@@ -55,38 +56,40 @@ impl Precompile for Blake2F {
...
+                .map_err(|e| ExitError::Other(e.to_string().into()))?;
         }
 
-        if input[212] != 0 && input[212] != 1 {
-            return Err(ExitError::Other(Borrowed("ERR_BLAKE2F_FINAL_FLAG")));
+        let flag = codec
+            .get_u8()
+            .map_err(|e| ExitError::Other(e.to_string().into()))?;
+
+        if flag != 0 && flag != 1 {
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply chain-evm/src/precompiles/blake2.rs:89]: Good point, but will need to determine what is the meaning of this consts to make a relevant naming 😊

**Comment by Mr-Leshiy** [comment]
@enzoc4 , yes currently I am trying to replace it with the cryptoxide usage in the different branch.

**Comment by Zeegomo** [comment]
I wonder whether `Codec` was supposed to provide a default way of serializing data (i.e.) allow different parts of the code to adopt the same standard for de-serialization without risking compatibility. With this new change instead we're back to every user having to decide which endianess to use, possibly mismatching it, or anyway not masking this detail which is not really relevant at all levels.
Maybe we can introduce something like a `DefaultCodec` that does this while allowing code reuse for things like this?

**Comment by Mr-Leshiy** [comment]
@Zeegomo, I have not catch your point, can you pls elaborate little bit more.
Do you mean that if we will modify `Codec` serialization, so we will need to distinguish these default serialization methods ?|
If so I think we can just extend the functionality of the `Codec` to provide a different way of serialization/deserialization data.

**Comment by Zeegomo** [comment]
What I mean is that `Codec` only had one possible way of serializing numbers (i.e. big endian), so you could not mess it up in different parts of the code (e.g. write `put_be_u64`  in serialization and `get_le_u64` when deserializing ). 

Also, the 'complexity' of endianess was masked away from the user, and all that was exposed was some standard way of serializing data. It's not really relevant when I want to serialize some domain-specific struct whether a number is big-endian or little-endian.
In addition, it removes some flexibility in the implementation of `Codec` as it's forced to use a specific format

**Comment by Zeegomo** [comment]
I'm just debating the level of abstraction of the tools used when serializing/deserializing, it's perfectly ok for `Codec` to allow multiple different choices. 

It feels to me like it's not important when I'm writing a `Deserialize` impl whether integers are big or little endian, and actually I don't care much about the specific encoding anyway (e.g. it could deserialize integers like CBOR to save space), provided it's sensible and consistent across the whole project.

`impl Deserialize for u64` could be interesting though.

**Comment by Mr-Leshiy** [comment]
@Zeegomo , @enzoc4  What do you think guys do we need to save these notes somewhere in the ticket or any other place, to not lose it ? I think it is a good discussion which should not just disappear 😊

**Comment by Zeegomo** [comment]
Personally, I do not dislike this kind of discussions directly here of GitHub if they are not super long, as the context is well defined ( slack works as well).
Basically, everything which is not Jira is ok 😅 , but we can always put a link to the actual discussion there.

**Pull request merged**